library/Vanilla/Embeds/CodePenEmbed.php

```diff
@@ -0,0 +1,96 @@
+<?php
+/**
+ * @copyright 2009-2018 Vanilla Forums Inc.
+ * @license https://opensource.org/licenses/GPL-2.0 GPL-2.0
+ */
+
+namespace Vanilla\Embeds;
+
+/**
+ * CodePen Embed.
+ */
+class CodePenEmbed extends Embed {
+
+    /** @inheritdoc */
+    protected $domains = ['codepen.io'];
+
+    /**
+     * CodePenEmbed constructor.
+     */
+    public function __construct() {
+        parent::__construct('codepen', 'image');
+    }
+
+    /**
+     * @inheritdoc
+     */
+    public function matchUrl(string $url) {
+        $data = [];
+        $oembedData = [];
+
+        if ($this->isNetworkEnabled()) {
+            $oembedData = $this->oembed("https://codepen.io/api/oembed?url=".urlencode($url)."&format=json");
+        }
+
+        if ($oembedData) {
+            if (array_key_exists('html', $oembedData)) {
+                $data['attributes'] = $this->parseHtml($oembedData['html']);
+            }
+            if (array_key_exists('height', $oembedData)) {
+                $data['height'] = $oembedData['height'];
+            }
+        }
+
+        return $data;
+    }
+
+    /**
+     * @inheritdoc
+     */
+    public function renderData(array $data): string {
+        $height = $data['height'] ?? '';
+        $embedUrl = $data['attributes']['embedUrl'] ?? '';
+        $style = $data['attributes']['style'] ?? '';
+        $id = $data['attributes']['id'];
+
+        $encodedHeight = htmlspecialchars($height);
+        $encodedEmbedUrl = htmlspecialchars($embedUrl);
+        $encodedStyle = htmlspecialchars($style);
```

Comment:
I'm not a big fan of allowing CodePen to insert arbitrary styles into Vanilla pages. Can you update `parseHtml` to only extract certain whitelisted styles?

Reply:
I think we mainly care about width and overflow, so I will parse those out.